### PR TITLE
Update to latest istio/api and implement validation for the Gateway

### PR DESCRIPTION
**Pull request by ZackButcher**
**What this PR does / why we need it**: Pull in the just-merged Gateway istio/api changes and implement validation for them. This PR does not wire up any code to produce Envoy config from the Gateway.

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pilot/model/validation_test.go [hunk 1]
```diff
@@ -1303,23 +1335,60 @@ var (
...
 	for _, c := range cases {
-		if got := ValidateMixerAttributes(c.in); (got == nil) != c.valid {
-			t.Errorf("ValidateMixerAttributes(%v): got(%v) != want(%v): %v", c.name, got == nil, c.valid, got)
-		}
+		t.Run(c.name, func(t *testing.T) {
+			attrs := &mpb.Attributes{
+				Attributes: map[string]*mpb.Attributes_AttributeValue{"key": c.in},
+			}
+			if got := ValidateMixerAttributes(attrs); (got == nil) != c.valid {
+				t.Errorf("ValidateMixerAttributes(%v): got(%v) != want(%v): %v", c.name, got == nil, c.valid, got)
```
Comment (ayj) [inline pilot/model/validation_test.go]: Is `ValidateMixerAttribute(%v)` and the explicit name needed here with subtests? I believe omitting these may give more concise test output, e.g.

```bash
--- FAIL: TestValidateMixerAttributes (0.00s)
    --- PASS: TestValidateMixerAttributes/happy duration (0.00s)
    --- FAIL: TestValidateMixerAttributes/nil duration (0.00s) 
    got(...) != want(...)   
```

File: pilot/model/validation.go [hunk 2]
```diff
@@ -36,11 +36,14 @@ import (
...
 	mccpb "istio.io/api/mixer/v1/config/client"
 	routing "istio.io/api/routing/v1alpha1"
+	routingv2 "istio.io/api/routing/v1alpha2"
 )
 
 const (
 	dns1123LabelMaxLength int    = 63
 	dns1123LabelFmt       string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+	// a wild-card prefix is an '*', a normal DNS1123 label with a leading '*' or '*-', or a normal DNS1123 label
+	wildcardPrefix string = `\*|(\*|\*-)?(` + dns1123LabelFmt + `)`
```
Comment (rshriram) [inline pilot/model/validation.go]: is this the same as envoy's virtual host implementation ?

File: pilot/model/validation.go [hunk 3]
```diff
@@ -234,22 +232,55 @@ func (t Labels) Validate() error {
...
+
+func ValidateWildcardDomain(domain string) error {
+	if err := checkDNS1123Preconditions(domain); err != nil {
+		return err
 	}
-	if len(fqdn) == 0 {
+	// We only allow wildcards in the first label; split off the first label (parts[0]) from the rest of the host (parts[1])
+	parts := strings.SplitN(domain, ".", 2)
+	if !IsWildcardDNS1123Label(parts[0]) {
+		return fmt.Errorf("domain name %q invalid (label %q invalid)", domain, parts[0])
```
Comment (rshriram) [inline pilot/model/validation.go]: what is Label here?

File: pilot/model/validation.go [hunk 4]
```diff
@@ -852,6 +883,70 @@ func ValidateIngressRule(msg proto.Message) error {
 	return errs
 }
 
+// ValidateGateway checks ingress rules
```
Comment (rshriram) [inline pilot/model/validation.go]: nit: checks gateway rules

File: pilot/model/validation_test.go [hunk 5]
```diff
@@ -1859,3 +1928,237 @@ func TestValidateEndUserAuthenticationPolicySpecBinding(t *testing.T) {
...
+			&routingv2.Gateway{
+				Servers: []*routingv2.Server{
+					{
+						Domains: []string{"foo.bar.com"},
+						Port:    &routingv2.Server_Port{Number: 7, Protocol: "http"},
+					},
+					{
+						Domains: []string{"baz.bar.com"},
+						Port:    &routingv2.Server_Port{Number: 66000, Protocol: "redis"},
+					}},
```
Comment (rshriram) [inline pilot/model/validation_test.go]: We were discussing about adding support for IPs/with or without CIDRs as well right? i.e. I should be able to specify foo.bar.com, 1.1.1.1, or service short name. I see tests for foo.bar.com, but not for IPs or service short name. 

That also reminds me that for service short names, we need to be able to interpolate the service's FQDN based on the calling pod's namespace (if this is kube). I think that can happen in another PR.

File: pilot/model/validation.go (hunk 3 above)
Comment (ZackButcher) [reply pilot/model/validation.go]: I just copied the wording from the original `ValidateFQDN` method, which uses the error string: `"domain name %q invalid (label %q invalid)"`. Happy to update if we want to use different language, though AFAIK RFC 1123 uses the language "label" to describe a dot-delineated sub-piece of a domain.

File: pilot/model/validation.go (hunk 4 above)
Comment (ZackButcher) [reply pilot/model/validation.go]: What do we want to validate here? Rules aren't directly embedded in the Gateway spec. We can check, when a gateway is created, that there are some route rules that reference it, but I think the better way would be to validate at RouteRule ingestion that the RouteRule references a valid Gateway. Either way I think we'll want to allow both Gateways with no RouteRules bound to them, and RouteRules that bind to non-extant Gateways.

File: pilot/model/validation.go (hunk 3 above)
Comment (rshriram) [reply pilot/model/validation.go]: Ah I see. I wanted clarification thats all. Odd term to use in the RFC though..

File: pilot/model/validation.go (hunk 4 above)
Comment (rshriram) [reply pilot/model/validation.go]: Sorry for the mis understanding..May be say "validategateway checks gateway configuration" ? rule is probably a wrong word to use..

File: pilot/model/validation.go (hunk 2 above)
Comment (rshriram) [reply pilot/model/validation.go]: yeah it is.. ignore comment

File: pilot/model/validation.go (hunk 4 above)
Comment (ZackButcher) [reply pilot/model/validation.go]: Oh whoops, sorry, totally misread your comment! Yes, comment string is a copy/paste error on my part, thanks.

File: pilot/model/validation_test.go (hunk 5 above)
Comment (ZackButcher) [reply pilot/model/validation_test.go]: I'll update PR to add validation for IPs and allow them to be used in the field too. Name canonicalization can come in a follow on PR as it's reasonably complex by itself.

**Pull request merged**